File: Last Stone Weight - Leetcode 1046/Last Stone Weight - Leetcode 1046.java
import java.util.PriorityQueue;

public class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> b - a);

        for (int stone : stones) {
            maxHeap.offer(stone);
        }

        while (maxHeap.size() > 1) {
            int largest = maxHeap.poll();
            int nextLargest = maxHeap.poll();

            if (largest != nextLargest) {
                maxHeap.offer(largest - nextLargest);
            }
        }

        return maxHeap.isEmpty() ? 0 : maxHeap.peek();
    }
}

File: Last Stone Weight - Leetcode 1046/Last Stone Weight - Leetcode 1046.cpp
#include <vector>
#include <queue>
#include <functional>

class Solution {
public:
    int lastStoneWeight(std::vector<int>& stones) {
        std::priority_queue<int> maxHeap(stones.begin(), stones.end());

        while (maxHeap.size() > 1) {
            int largest = maxHeap.top();
            maxHeap.pop();
            int nextLargest = maxHeap.top();
            maxHeap.pop();

            if (largest != nextLargest) {
                maxHeap.push(largest - nextLargest);
            }
        }

        return maxHeap.empty() ? 0 : maxHeap.top();
    }
};
